src/main/java/bio/terra/service/profile/ProfileService.java

```diff
@@ -232,4 +250,49 @@ public void verifyAccount(BillingProfileRequestModel request, AuthenticatedUserR
         }
     }
 
+    // -- profile upgrade --
+
+    // Billing profiles prior to the introduction of this service did not have sam resources behind them.
+    // This code generates sam resources for any billing profile it does not have one. It sets the
+    // stewards group as the owner of the billing profile.
+    //
+    // In the current state, no billing profiles have sam resources, so the resources list will be empty.
+    // When we upgrade a billing profile, we give stewards owner role. Since we run this
+    // as a steward, we will be able to see any profiles we have already converted.
+    //
+    // The hole here is if someone gets in and creates a billing profile with a resource before
+    // we run the upgrade. We would not retrieve that one here. However, Sam would not let us create
+    // the resource in the step below. We would log an error, but otherwise, things would work.
+
+    public UpgradeResponseModel upgradeProfileResources(UpgradeModel request, AuthenticatedUserRequest user) {
+
+        Instant startTime = Instant.now();
+        List<BillingProfileModel> profiles = profileDao.getOldBillingProfiles();
+
+        List<UUID> resources = iamService.listAuthorizedResources(user, IamResourceType.SPEND_PROFILE);
+        Set<String> profileResourceSet;
+        if (resources == null) {
+            profileResourceSet = Collections.EMPTY_SET;
+        } else {
+            profileResourceSet = resources.stream().map(r -> toString()).collect(Collectors.toSet());
+        }
+
+        for (BillingProfileModel profile : profiles) {
+            String profileId = profile.getId();
+            if (!profileResourceSet.contains(profileId)) {
+                try {
+                    // No Sam profile, so let's make one - stewards are automatically added as owner
+                    logger.info("Creating profile resource for {} ({})", profile.getProfileName(), profileId);
+                    iamService.createProfileResource(user, profile.getId());
+                } catch (Exception ex) {
+                    logger.error("IAM failure during upgrade", ex);
```

Comment:
Is this correct? Wouldn't we want to propagate the exception in this case?

Reply:
I don't think so. If there is some oddball state in Sam, I don't want the rest of the upgrade to faill. We can use the log to see what we need to manually fix.

Reply:
OK. We would need to be careful to make sure that error logs are preserved during an update procedure, to avoid data loss. I'm thinking about the general case for running updates as well as this specific case.